### ISSUE TITLE
vcpkg Support for building on Windows

### ISSUE THREAD

**Issue by angelmixu** [libvips/libvips#868]
Hi!

Just stumbled onto vcpkg dependency manager: https://github.com/Microsoft/vcpkg
And became amazed of all the libraries being supported nowadays.

I was wondering if a libvips port could be created for vcpkg.
The building process for Visual Studio would be quite straight forward, I've been testing it with lots of dependencies and works very well. Also, most of libvips dependencies are supported.

I know there's an already stablished build process using MSYS and MinGW, but this one would be for targetting directly Visual Studio.

I opened an issue on vcpkg for if someone could make the port: https://github.com/Microsoft/vcpkg/issues/2676
Maybe I can give some help :)

What do you think?

**Comment by jcupitt** [comment]
Sure, happy to help if I can.

**Comment by jcupitt** [comment]
Actually the current build system does not use MSYS and uses the mingw backend of standard gcc. 

You install docker, then it's one-line command to download, patch, configure, build, install all the 40 (!!) projects that the libvips binary needs, and package it up as a nice zip.

I agree compiling with MSVC is difficult :(

**Comment by angelmixu** [comment]
Oh ok!

Looking a little on documentation and some ports it looks like they mostly use CMake ( sometimes with Ninja ) for building libraries. Maybe we should first try to create a CMake project for building.
Apart from that, I've found this: https://vcpkg.readthedocs.io/en/latest/specifications/export-command/
A command which purpose is to make a package with binaries and its dependencies, a zip with all needed stuff.

At this moment I'm quite busy and I haven't used CMake ever so I should learn it some day.
I hope in some weeks to be not so busy so I could take a better look at it!

**Comment by jcupitt** [comment]
Switching libvips to cmake would be a big project! Perhaps it will need to happen in the next year or two, we'll see. 

I'm not keen to make the official libvips binaries on Windows. I don't have easy access to a win machine, so it would be very inconvenient for me. I'm keen to support a vcpkg libvips package though.

**Comment by angelmixu** [comment]
No problem, when I have some time I can take a look!

I'm also a Linux user, but I have to work on lots of platforms :P

**Comment by jcupitt** [comment]
Hello @leha-bot, it's kind of you to offer.

Sure, let's make a test branch with a cmake build system and evaluate it. It would need to support cross-compiling from linux to win, native linux builds, native macos builds, and native win builds. 

As long as it's not too horrible, we'll switch for 8.10.

@kleisauke and @lovell, do you have any feelings about this, or possible alternatives? meson, maybe? I find cmake a bit ugly myself (poor scripting language, clumsy command-line, cross-compiling is awkward, poor support for pkg-config), but perhaps there's no real alternative.

**Comment by lovell** [comment]
In my experience cmake is more mature and therefore less painful than meson when it comes to cross-compiling.

Using cmake places a build-time dependency on the consumer of the release tarballs to also have cmake installed, which is usually not a problem.

As long as pkg-config files are still available I'm happy.

**Comment by angelmixu** [comment]
Hey! Count with me if you need any help!
I have migrated the dependency management at my job to vcpkg and we're quite happy with it :)

**Comment by kleisauke** [comment]
I'm fine with adding a CMake build system to ease the vcpkg packaging. I just checked to see if this hasn't been done before. Looks like someone already put a lot of work into this:
https://github.com/libvips/libvips/compare/master...wingc-china:master

We should probably not remove the autotools build system after we've integrated this. I think the Fedora and Debian package maintainers still prefer autotools over CMake (for e.g. libwebp, which has support for building with CMake, is built with autotools on these distributions).

fwiw, there's already a Fedora (and CentOS) spec file available:
https://src.fedoraproject.org/rpms/vips
https://git.remirepo.net/cgit/rpms/lib/vips.git/

**Comment by angelmixu** [comment]
I was taking a look at changes of @wingc-china and looks like he/she fixed some stuff that we have already fixed (and also some issues I had with my custom Visual Studio solution).
I can see in that CMakeFileLists.txt there are all .c files listed, wouldn't it be a better solution to glob all .c files, thus, if there's some files added or removed, the script would have to be changed.

So, how should we proceed with this?
